File: apple/Sources/R2CCore/TrackerTabletLink.swift
import CryptoKit
import Foundation

public enum TrackerTabletLink {
    private static let codeDigestBytes = 4

    public static func organizationDesignator(from trackerURLPrefix: String) -> String? {
        let trimmed = trackerURLPrefix.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let components = URLComponents(string: trimmed) else { return nil }
        return components.path
            .split(separator: "/")
            .last
            .map { String($0).lowercased() }
    }

    public static func shortURL(
        trackerURLPrefix: String,
        tabletName: String
    ) -> URL? {
        let trimmedPrefix = trackerURLPrefix.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let organization = organizationDesignator(from: trimmedPrefix),
              !normalize(tabletName).isEmpty,
              var components = URLComponents(string: trimmedPrefix),
              components.scheme != nil,
              components.host != nil
        else { return nil }
        components.path = "/t/" + code(
            organization: organization,
            tabletName: tabletName
        )
        components.query = nil
        components.fragment = nil
        return components.url
    }

    public static func markerDescription(
        trackerURLPrefix: String,
        tabletName: String,
        trackerConnected: Bool = true
    ) -> String {
        guard trackerConnected else { return "" }
        guard let url = shortURL(
            trackerURLPrefix: trackerURLPrefix,
            tabletName: tabletName
        ) else { return "" }
        return url.absoluteString
    }

    public static func streamShortURL(
        trackerURLPrefix: String,
        tabletName: String,
        videoStream: String
    ) -> URL? {
        let trimmedPrefix = trackerURLPrefix.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let organization = organizationDesignator(from: trimmedPrefix),
              !normalize(tabletName).isEmpty,
              !normalize(videoStream).isEmpty,
              var components = URLComponents(string: trimmedPrefix),
              components.scheme != nil,
              components.host != nil
        else { return nil }
        components.path = "/s/" + streamCode(
            organization: organization,
            tabletName: tabletName,
            videoStream: videoStream
        )
        components.query = nil
        components.fragment = nil
        return components.url
    }

    public static func recordingShortURL(
        trackerURLPrefix: String,
        tabletName: String,
        sessionID: String
    ) -> URL? {
        let trimmedPrefix = trackerURLPrefix.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let organization = organizationDesignator(from: trimmedPrefix),
              !normalize(tabletName).isEmpty,
              !normalize(sessionID).isEmpty,
              var components = URLComponents(string: trimmedPrefix),
              components.scheme != nil,
              components.host != nil
        else { return nil }
        components.path = "/v/" + recordingCode(
            organization: organization,
            tabletName: tabletName,
            sessionID: sessionID
        )
        components.query = nil
        components.fragment = nil
        return components.url
    }

    public static func thumbnailURL(
        trackerURLPrefix: String,
        tabletName: String,
        streamSessionID: String,
        thumbnailRevision: String
    ) -> URL? {
        let trimmedPrefix = trackerURLPrefix.trimmingCharacters(in: .whitespacesAndNewlines)
        let sessionID = streamSessionID.trimmingCharacters(in: .whitespacesAndNewlines)
        let revision = thumbnailRevision.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let organization = organizationDesignator(from: trimmedPrefix),
              !normalize(tabletName).isEmpty,
              !sessionID.isEmpty,
              !revision.isEmpty,
              var components = URLComponents(string: trimmedPrefix),
              components.scheme != nil,
              components.host != nil
        else { return nil }
        let tabletCode = code(organization: organization, tabletName: tabletName)
        components.path = "/r2c-thumbnail/\(tabletCode)/\(sessionID).jpg"
        components.queryItems = [URLQueryItem(name: "timestamp", value: revision)]
        components.fragment = nil
        return components.url
    }

    public static func code(
        organization: String,
        tabletName: String
    ) -> String {
        let material = "/\(normalize(organization))/streams/\(normalize(tabletName))"
        let digest = SHA256.hash(data: Data(material.utf8))
        return Data(digest.prefix(codeDigestBytes))
            .base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    public static func streamCode(
        organization: String,
        tabletName: String,
        videoStream: String
    ) -> String {
        let material = "/\(normalize(organization))/streams/\(normalize(tabletName))/\(normalize(videoStream))"
        let digest = SHA256.hash(data: Data(material.utf8))
        return Data(digest.prefix(codeDigestBytes))
            .base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    public static func recordingCode(
        organization: String,
        tabletName: String,
        sessionID: String
    ) -> String {
        let material = "/\(normalize(organization))/streams/\(normalize(tabletName))/session/\(normalize(sessionID))"
        let digest = SHA256.hash(data: Data(material.utf8))
        return Data(digest.prefix(codeDigestBytes))
            .base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    private static func normalize(_ value: String) -> String {
        value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    }
}

File: app/src/main/java/org/ncssar/rid2caltopo/data/TrackerTabletLink.java
package org.ncssar.rid2caltopo.data;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Base64;
import java.util.Locale;

/** Shared contract for the tracker tablet link published in a CalTopo device marker. */
public final class TrackerTabletLink {
    private static final int CODE_DIGEST_BYTES = 4;

    private TrackerTabletLink() {
    }

    @Nullable
    public static String organizationDesignator(@Nullable String trackerUrlPrefix) {
        String value = trackerUrlPrefix == null ? "" : trackerUrlPrefix.trim();
        if (value.isEmpty()) return null;
        try {
            URI uri = URI.create(value);
            String path = uri.getPath() == null ? "" : uri.getPath();
            String[] segments = path.split("/");
            for (int index = segments.length - 1; index >= 0; index--) {
                String segment = segments[index].trim();
                if (!segment.isEmpty()) return segment.toLowerCase(Locale.ROOT);
            }
        } catch (IllegalArgumentException ignored) {
            return null;
        }
        return null;
    }

    @Nullable
    public static String shortUrl(
            @Nullable String trackerUrlPrefix,
            @Nullable String tabletName
    ) {
        String organization = organizationDesignator(trackerUrlPrefix);
        String tablet = normalize(tabletName);
        if (organization == null || tablet.isEmpty()) return null;
        try {
            URI tracker = URI.create(trackerUrlPrefix.trim());
            if (tracker.getScheme() == null || tracker.getRawAuthority() == null) return null;
            String code = code(organization, tablet);
            return new URI(
                    tracker.getScheme(),
                    tracker.getRawAuthority(),
                    "/t/" + code,
                    null,
                    null
            ).toASCIIString();
        } catch (Exception ignored) {
            return null;
        }
    }

    @NonNull
    public static String markerDescription(
            @Nullable String trackerUrlPrefix,
            @Nullable String tabletName
    ) {
        return markerDescription(trackerUrlPrefix, tabletName, true);
    }

    @NonNull
    public static String markerDescription(
            @Nullable String trackerUrlPrefix,
            @Nullable String tabletName,
            boolean trackerConnected
    ) {
        if (!trackerConnected) return "";
        String url = shortUrl(trackerUrlPrefix, tabletName);
        return url == null ? "" : url;
    }

    @Nullable
    public static String streamShortUrl(
            @Nullable String trackerUrlPrefix,
            @Nullable String tabletName,
            @Nullable String videoStream
    ) {
        String organization = organizationDesignator(trackerUrlPrefix);
        String tablet = normalize(tabletName);
        String stream = normalize(videoStream);
        if (organization == null || tablet.isEmpty() || stream.isEmpty()) return null;
        try {
            URI tracker = URI.create(trackerUrlPrefix.trim());
            if (tracker.getScheme() == null || tracker.getRawAuthority() == null) return null;
            return new URI(
                    tracker.getScheme(),
                    tracker.getRawAuthority(),
                    "/s/" + streamCode(organization, tablet, stream),
                    null,
                    null
            ).toASCIIString();
        } catch (Exception ignored) {
            return null;
        }
    }

    @Nullable
    public static String recordingShortUrl(
            @Nullable String trackerUrlPrefix,
            @Nullable String tabletName,
            @Nullable String sessionId
    ) {
        String organization = organizationDesignator(trackerUrlPrefix);
        String tablet = normalize(tabletName);
        String session = normalize(sessionId);
        if (organization == null || tablet.isEmpty() || session.isEmpty()) return null;
        try {
            URI tracker = URI.create(trackerUrlPrefix.trim());
            if (tracker.getScheme() == null || tracker.getRawAuthority() == null) return null;
            return new URI(
                    tracker.getScheme(),
                    tracker.getRawAuthority(),
                    "/v/" + recordingCode(organization, tablet, session),
                    null,
                    null
            ).toASCIIString();
        } catch (Exception ignored) {
            return null;
        }
    }

    @Nullable
    public static String thumbnailUrl(
            @Nullable String trackerUrlPrefix,
            @Nullable String tabletName,
            @Nullable String streamSessionId
    ) {
        String organization = organizationDesignator(trackerUrlPrefix);
        String tablet = normalize(tabletName);
        String sessionId = streamSessionId == null ? "" : streamSessionId.trim();
        if (organization == null || tablet.isEmpty() || sessionId.isEmpty()) return null;
        try {
            URI tracker = URI.create(trackerUrlPrefix.trim());
            if (tracker.getScheme() == null || tracker.getRawAuthority() == null) return null;
            return new URI(
                    tracker.getScheme(),
                    tracker.getRawAuthority(),
                    "/r2c-thumbnail/" + code(organization, tablet) + "/" + sessionId + ".jpg",
                    null,
                    null
            ).toASCIIString();
        } catch (Exception ignored) {
            return null;
        }
    }

    @NonNull
    static String code(
            @NonNull String organization,
            @NonNull String tabletName
    ) throws Exception {
        String material = "/" + normalize(organization)
                + "/streams/" + normalize(tabletName);
        byte[] digest = MessageDigest.getInstance("SHA-256")
                .digest(material.getBytes(StandardCharsets.UTF_8));
        byte[] prefix = new byte[CODE_DIGEST_BYTES];
        System.arraycopy(digest, 0, prefix, 0, CODE_DIGEST_BYTES);
        return Base64.getUrlEncoder().withoutPadding().encodeToString(prefix);
    }

    @NonNull
    static String streamCode(
            @NonNull String organization,
            @NonNull String tabletName,
            @NonNull String videoStream
    ) throws Exception {
        String material = "/" + normalize(organization)
                + "/streams/" + normalize(tabletName)
                + "/" + normalize(videoStream);
        byte[] digest = MessageDigest.getInstance("SHA-256")
                .digest(material.getBytes(StandardCharsets.UTF_8));
        byte[] prefix = new byte[CODE_DIGEST_BYTES];
        System.arraycopy(digest, 0, prefix, 0, CODE_DIGEST_BYTES);
        return Base64.getUrlEncoder().withoutPadding().encodeToString(prefix);
    }

    @NonNull
    static String recordingCode(
            @NonNull String organization,
            @NonNull String tabletName,
            @NonNull String sessionId
    ) throws Exception {
        String material = "/" + normalize(organization)
                + "/streams/" + normalize(tabletName)
                + "/session/" + normalize(sessionId);
        byte[] digest = MessageDigest.getInstance("SHA-256")
                .digest(material.getBytes(StandardCharsets.UTF_8));
        byte[] prefix = new byte[CODE_DIGEST_BYTES];
        System.arraycopy(digest, 0, prefix, 0, CODE_DIGEST_BYTES);
        return Base64.getUrlEncoder().withoutPadding().encodeToString(prefix);
    }

    @NonNull
    private static String normalize(@Nullable String value) {
        return value == null ? "" : value.trim().toLowerCase(Locale.ROOT);
    }
}
